### PR TITLE
Add token_creation_wrapper around token creation

### PR DESCRIPTION
**Pull request by ryansch**
### Summary

Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together.

### Other Information

If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating CHANGELOG.md file or are asked to update it by reviewers,
please add the changelog entry at the top of the file.

Thanks for contributing to Doorkeeper project!

### CODE REVIEW

**Comment by ryansch** [comment]
@nbulaj Whoa. Did you mean to force push to `master`?

**Comment by nbulaj** [comment]
Uh, yes. You just created this PR at the "wrong time" :D Sorry for that

**Comment by ryansch** [comment]
No worries then, I'll force push as well.

File: lib/doorkeeper/models/access_token_mixin.rb [hunk 1]
```diff
@@ -177,27 +177,39 @@ def scopes_match?(token_scopes, param_scopes, app_scopes)
       #
       # @return [Doorkeeper::AccessToken] existing record or a new one
       #
-      def find_or_create_for(application, resource_owner, scopes, expires_in, use_refresh_token)
-        if Doorkeeper.config.reuse_access_token
-          access_token = matching_token_for(application, resource_owner, scopes)
-
-          return access_token if access_token&.reusable?
-        end
+      def find_or_create_for(application, resource_owner, scopes, expires_in, use_refresh_token, token_attributes = {})
```
[the remark this answers is not in the document]
Comment (ryansch) [reply lib/doorkeeper/models/access_token_mixin.rb]: I can break that up into multiple lines but what it really needs is a refactor to named params.

File: lib/doorkeeper/models/access_token_mixin.rb [hunk 2]
```diff
@@ -177,27 +177,46 @@ def scopes_match?(token_scopes, param_scopes, app_scopes)
...
```
[the remark this answers is not in the document]
Comment (ryansch) [reply lib/doorkeeper/models/access_token_mixin.rb]: I can break that up into multiple lines but what it really needs is a refactor to named params.

File: spec/requests/flows/authorization_code_spec.rb [hunk 3]
```diff
@@ -526,5 +526,37 @@ def authorize(redirect_url)
       should_have_json "error", "invalid_grant"
       should_have_json "error_description", translated_error_message(:invalid_grant)
     end
+
+    context 'with a token_creation_wrapper' do
+      it 'wraps token creation' do
+        # Creating the wrapper here so we have access to wrapper_count
+        wrapper_count = 0
+        wrapper = ->(&block) do
```
[the remark this answers is not in the document]
Comment (ryansch) [reply spec/requests/flows/authorization_code_spec.rb]: That's a hard disagree from me.

**Comment by nbulaj** [comment]
Before you finish - why don't you want to just use hooks or `before_create` callback on required model to add any attributes you need? For now this PR looks like the feature you need

**Comment by ryansch** [comment]
@nbulaj Our use case is actually going to be putting a high performance distributed lock around the creation of tokens. We have an api integrator that sends us many concurrent duplicate requests. I'm going to coalesce then into one token creation.

**Comment by ryansch** [comment]
Btw, this sort of comment is exactly why I opened this as a draft PR! 😄

**Comment by nbulaj** [comment]
This loos like something that I don';t sure must be implemented in gem core :( I see a lot of requests when somebody needs something for this project but we can't merge every custom feature in the gem. Instead I like the changes when we just make our internals more extendable **without** need to change the gem itself (so everybody can just create controller, or set custom methods on models and so on - just like Devise do and many other gems)(

Can't you just set your custom AccessToken model (that Doorkeeper will use) and override `.find_or_create_for` method for your requirements?

**Comment by ryansch** [comment]
@nbulaj I sure can! I wanted to see if anyone thought it would be useful outside of our use case before I took it closed source.

One of the things I think is useful apart from the wrapper changes is making the refresh token creation go through `find_or_create_for` like all of the other flows. Would you entertain a PR with just that part?

**Pull request closed without merging**